tests/base/vectorization_17.cc

```diff
@@ -0,0 +1,108 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2015 - 2020 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// test mixed precision arithmetic for vectorized array
+
+#include <deal.II/base/vectorization.h>
+
+#include "../tests.h"
+
+
+template <typename VectorizedArrayType>
+void
+print(const std::string &msg, const VectorizedArrayType &array)
+{
+  deallog << msg << "  ";
+  for (unsigned int i = 0; i < VectorizedArrayType::size(); ++i)
+    deallog << array[i] << ' ';
+  deallog << std::endl;
+}
+
+
+template <typename VectorizedArrayType, typename NumberType2>
+void
+do_test(const VectorizedArrayType array, const NumberType2 number)
+{
+  deallog << "  test " << VectorizedArrayType::size() << " array elements"
+          << std::endl;
+
+  const auto addition_1 = array + number;
+  const auto addition_2 = number + array;
+
+  print("add 1", addition_1);
+  print("add 2", addition_2);
+
+  const auto subtraction_1 = array - number;
+  const auto subtraction_2 = number - array;
+
+  print("sub 1", subtraction_1);
+  print("sub 2", subtraction_2);
+
+  const auto multiplication_1 = array * number;
+  const auto multiplication_2 = number * array;
+
+  print("mult 1", multiplication_1);
+  print("mult 2", multiplication_2);
+
+  const auto division_1 = array / number;
+  const auto division_2 = number / array;
+
+  print("div 1", division_1);
+  print("div 2", division_2);
+}
+
+
+int
+main()
+{
+  initlog();
+
+#if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 512
+  do_test(VectorizedArray<double, 8>(2.0), 3.0f);
+  do_test(VectorizedArray<float, 16>(2.0), 3.0);
+
+  do_test(VectorizedArray<double, 8>(2.0), VectorizedArray<float, 16>(3.0));
```

Comment:
Hi @kronbichler. I've managed to do more or less what I hoped to with this PR. I've added some more generic binary math operators that are compatible with two different vectorisation types. I was initially targeting, say, supporting `VecArray<double>()*VecArray<complex<double>>()` but as a consequence of the current implementation this situation is valid. Do you think that this might ever be useful? From the [test output](https://github.com/dealii/dealii/pull/13537/files#diff-3d5448a1eeaa9d6b88877a56105d7a5c7642842771a75cb84ba9e17d9a62453aR21-R28) I'd worry that this is probably more confusing than anything else.

Reply:
I forgot about the case of `width==1`. I've reminded myself that I gave a small motivation [here](https://github.com/dealii/dealii/pull/13537/files#diff-dcb6223036ccab5f3890a70aeeb8300e8faaaf21c0cc0c63c41ebe7f7a8906a6R4747-R4787), but I'd still like to hear whether or not you think this could be a reasonable thing to do or if its simply confusing.